rfd/0203-database-healthchecks.md

```diff
@@ -0,0 +1,611 @@
+
+For comparison, here is a table with the default settings for each:
+
+| Name                           | Interval | Timeout              |    Healthy Threshold |  Unhealthy Threshold |
+| ------------------------------ | -------- | -------------------- | -------------------- | -------------------- |
+| AWS NLB Target Groups [^2]     | 30s      | 10s                  |                    5 |                    2 |
+| Azure [^3][^4]                 | 5s       | 5s                   | (not configurable) 1 | (not configurable) 1 |
+| GCP [^5]                       | 5s       | 5s                   |                    2 |                    2 |
+| Kube Readiness Probe [^6]      | 10s      | 1s                   |                    1 |                    3 |
+| Docker [^7]                    | 30s      | 30s                  | (not configurable) 1 |                    3 |
+| Teleport                       | 10s      | 5s                   |                    2 |                    1 |
+
+As a brief summary of the Teleport health checker behavior:
+
+1. it does the first check immediately
+2. the first heatlh check will always transition to `healthy` or `unhealthy` regardless of configured thresholds
+3. it repeats the check after interval time
+4. each check blocks until pass or fail
+
+For Teleport the minimum time initialized->healthy or initialized->unhealthy is roughly 0s.
+
+The maximum time is bounded by the timeout: 5s.
+
+Adding the 5s heartbeat polling period to the max times, we get a max time to broadcast the status change: 10s.
+
+Thus, broadcasting the health status for a new or updated resource will take between 0s and 10s.
+
+This is fast enough for web UI enrollment interactions that might wait for the resource network status to be reported.
+
+The default healthy threshold is set to a higher value than the unhealthy threshold to help avoid false positive `healthy` status when network connectivity is unreliable.
+It makes sense to bias the health status to `unhealthy` when the network is unreliable anyway.
+
+[^2]: https://docs.aws.amazon.com/elasticloadbalancing/latest/network/target-group-health-checks.html
+[^3]: https://learn.microsoft.com/en-us/azure/load-balancer/load-balancer-custom-probe-overview
+[^4]: Azure timeouts always match the interval, they cannot be configured independently.
+[^5]: https://docs.docker.com/reference/dockerfile/#healthcheck
+[^6]: https://cloud.google.com/load-balancing/docs/health-check-concepts
+[^7]: https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#configure-probes
+
+#### Config example
+
+    $ tctl get health_check_config/default
+    kind: health_check_config
+    version: v3
+    metadata:
+      name: "default"
+      labels:
+        teleport.internal/resource-type: preset
```

Comment:
Does this mean this will get overridden on each auth restart? Idk, feels like we should just let users tweak it however they want (unlike, say, preset roles).

Reply:
`teleport.internal/resource-type: preset` roles can be customized by users, we have a fairly complex rule addition logic, but basically Teleport will add new rules while keeping the user customization.

`teleport.internal/resource-type: system` is the annotation for role that are force-applied on every startup.